### ISSUE TITLE
Repo: Convert `nullThrows` to an assertion function

### ISSUE THREAD

**Issue by kirkwaiblinger** [typescript-eslint/typescript-eslint#10077]
### Suggestion

Right now our `nullThrows` utility has signature like so:

```ts
declare function nullThrows<T>(x: T, reason: string): NonNullable<T>;
```

which basically models a runtime non-null assertion operator `x!`. 

However, _unlike_ the non-null assertion operator, we don't have a way to flag when the argument provided is already non-null

```ts
const nonNull = 'foo';
const foo1 = nonNull!; // violates no-unnecessary-type-assertion
const foo2 = nullThrows(foo1, 'reason'); // no linter report.
```

We could pretty easily right our own custom rule in order to flag this scenario, however, now that https://github.com/typescript-eslint/typescript-eslint/pull/10009 has landed, I propose let's change the signature to be an assertion function so that we can dogfood no-unnecessary-condition's `checkTypePredicates` functionality.

```ts
declare function nullThrows<T>(x: T, reason: string): asserts x is NonNullable<T>;

const nonNull = 'foo';

nullThrows(nonNull, 'reason'); // no-unnecessary-condition - types are already equal
```

**Comment by bradzacher** [comment]
This was done on purpose so that you can use it as an expression. You can't use an assertion function in that way.

Additionally assertion functions have rules around how they must be imported to ensure TS can correctly use them, but it's current signature can be imported in any form.

**Comment by kirkwaiblinger** [comment]
> This was done on purpose so that you can use it as an expression. You can't use an assertion function in that way.

I figured, right, but is there a reason that we need it to be used as an expression? Just curious.

**Comment by bradzacher** [comment]
IMO it's much nicer to write:
```ts
const property = nullThrows(node.property);
```
Rather than:
```ts
nullThrows(node.property);
const property = node.property;
```

IMO it's more ergonomic to do the assertion at the point of access, rather than pre-asserting some state. Same reason that people use non-null assertions and optional chains.

**Comment by kirkwaiblinger** [comment]
In principle, I agree, but I think that this is basically negated in the actual implementation due to the reason string being required. 

That is to say, I really don't think 

https://github.com/typescript-eslint/typescript-eslint/blob/d0e35d9897221896a80e33f5a0f96ec138a8911c/packages/eslint-plugin/src/rules/consistent-type-exports.ts#L173-L188

is any better than 

```ts
          fix(fixer) {
            const asteriskToken = context.sourceCode.getFirstToken(
              node,
              token =>
                token.type === AST_TOKEN_TYPES.Punctuator &&
                token.value === '*',
            );

            assertNotNullish(
              asteriskToken,
              NullThrowsReasons.MissingToken(
                'asterisk',
                'export all declaration',
              ),
            );

            return fixer.insertTextBefore(asteriskToken, 'type ');
          },
```

Actually, I think the second is much more understandable.

***

In any case, as mentioned, the real motivation for changing this would be to dogfood more cases for the no-unnecessary-condition check to ourselves. So as long as we consider the readability roughly equal, I'd be in favor of switching to an assertion here. But it's not really an important thing either, so I don't feel strongly.
